rewrite-core/src/main/java/org/openrewrite/rpc/request/Visit.java

```diff
@@ -0,0 +1,152 @@
+ */
+package org.openrewrite.rpc.request;
+
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.moderne.jsonrpc.JsonRpcMethod;
+import lombok.RequiredArgsConstructor;
+import lombok.Value;
+import org.jspecify.annotations.Nullable;
+import org.openrewrite.*;
+import org.openrewrite.internal.ObjectMappers;
+import org.openrewrite.scheduling.RecipeRunCycle;
+import org.openrewrite.scheduling.WatchableExecutionContext;
+import org.openrewrite.table.RecipeRunStats;
+import org.openrewrite.table.SourcesFileErrors;
+import org.openrewrite.table.SourcesFileResults;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+
+@Value
+public class Visit implements RpcRequest {
+    String visitor;
+
+    @Nullable
+    Map<String, Object> visitorOptions;
+
+    String treeId;
+
+    /**
+     * An ID of the p value stored in the caller's local object cache.
+     */
+    String p;
+
+    /**
+     * A list of IDs representing the cursor whose objects are stored in the
+     * caller's local object cache.
+     */
+    @Nullable
+    List<String> cursor;
+
+    @RequiredArgsConstructor
+    public static class Handler extends JsonRpcMethod<Visit> {
+        private static final ObjectMapper mapper = ObjectMappers.propertyBasedMapper(null);
+
+        private final Map<String, Object> remoteObjects;
+        private final Map<String, Object> localObjects;
+        private final Map<String, Recipe> preparedRecipes;
+        private final Function<String, ?> getObject;
+        private final Function<@Nullable List<String>, Cursor> getCursor;
+
+        @Override
+        protected Object handle(Visit request) throws Exception {
+            //noinspection unchecked
+            TreeVisitor<?, Object> visitor = (TreeVisitor<?, Object>) instantiateVisitor(request);
+
+            SourceFile before = (SourceFile) getObject.apply(request.getTreeId());
+            localObjects.put(before.getId().toString(), before);
+
+            Object p = getVisitorP(request);
+
+            SourceFile after = (SourceFile) visitor.visit(before, p, getCursor.apply(
+                    request.getCursor()));
+            if (after == null) {
+                localObjects.remove(before.getId().toString());
+            } else {
+                localObjects.put(after.getId().toString(), after);
+            }
+
+            maybeUpdateExecutionContext(request.getP(), p);
+
+            return new VisitResponse(before != after);
+        }
+
+        private TreeVisitor<?, ?> instantiateVisitor(Visit request) {
+            String visitorName = request.getVisitor();
+
+            if (visitorName.startsWith("scan:")) {
+                //noinspection unchecked
+                ScanningRecipe<Object> recipe = (ScanningRecipe<Object>) preparedRecipes.get(visitorName.substring("scan:".length()));
+                return recipe.getScanner(0);
```

Comment:
Unless we already want to instantiate and register the accumulator as part of the `PrepareRecipe` requests, we need to do it lazily at this point at the latest. This means that we in `handle()` need to call `getVisitorP()` first and pass it in here and then check if the accumulator has already been registered in the execution context.

Reply:
I don't understand the concern here. ScanningRecipe already initializes the accumulator lazily at the point where it supplies it to the visitor/scanner.

<img width="894" alt="image" src="https://github.com/user-attachments/assets/e4170a34-9f32-4bca-9e87-466fe88f4b9f" />